CMakeIntegratedOpenCL.cmake

```diff
@@ -0,0 +1,88 @@
+set(BUILD_SHARED_LIBS OFF CACHE BOOL "" FORCE)
+set(BOOST_VERSION_DOT "1.74")
+string(REPLACE "." "_" BOOST_VERSION_UNDERSCORE ${BOOST_VERSION_DOT})
+
+set(OPENCL_HEADER_REPOSITORY "https://github.com/KhronosGroup/OpenCL-Headers.git")
+set(OPENCL_HEADER_TAG "1b2a1850f410aaaaeaa56cead5a179b5aea4918e")
+
+set(OPENCL_LOADER_REPOSITORY "https://github.com/KhronosGroup/OpenCL-ICD-Loader.git")
+set(OPENCL_LOADER_TAG "862eebe7ca733c398334a8db8481172a7d3a3c47")
+
+set(BOOST_REPOSITORY "https://github.com/boostorg/boost.git")
+set(BOOST_TAG "boost-${BOOST_VERSION_DOT}.0")
+execute_process(COMMAND git rev-parse HEAD WORKING_DIRECTORY ${PROJECT_SOURCE_DIR} OUTPUT_VARIABLE LIGHTGBM_TAG OUTPUT_STRIP_TRAILING_WHITESPACE ERROR_QUIET)
+string(SHA1 INTEGRATED_OPENCL_STAMP "${OPENCL_HEADER_REPOSITORY}@${OPENCL_HEADER_TAG};${OPENCL_LOADER_REPOSITORY}@${OPENCL_LOADER_TAG};${BOOST_REPOSITORY}@${BOOST_TAG};lightgbm@${LIGHTGBM_TAG}")
+message(STATUS "Integrated OpenCL build stamp: ${INTEGRATED_OPENCL_STAMP}")
+
+# Build Independent OpenCL library
+include(FetchContent)
+FetchContent_Declare(OpenCL-Headers GIT_REPOSITORY ${OPENCL_HEADER_REPOSITORY} GIT_TAG ${OPENCL_HEADER_TAG})
+FetchContent_GetProperties(OpenCL-Headers)
+if(NOT OpenCL-Headers_POPULATED)
+  FetchContent_Populate(OpenCL-Headers)
+  message(STATUS "Populated OpenCL Headers")
+endif()
+set(OPENCL_ICD_LOADER_HEADERS_DIR ${opencl-headers_SOURCE_DIR} CACHE PATH "") # for OpenCL ICD Loader
+set(OpenCL_INCLUDE_DIR ${opencl-headers_SOURCE_DIR} CACHE PATH "") # for Boost::Compute
+
+FetchContent_Declare(OpenCL-ICD-Loader GIT_REPOSITORY ${OPENCL_LOADER_REPOSITORY} GIT_TAG ${OPENCL_LOADER_TAG})
+FetchContent_GetProperties(OpenCL-ICD-Loader)
+if(NOT OpenCL-ICD-Loader_POPULATED)
+  FetchContent_Populate(OpenCL-ICD-Loader)
+  set(USE_DYNAMIC_VCXX_RUNTIME ON)
+  add_subdirectory(${opencl-icd-loader_SOURCE_DIR} ${opencl-icd-loader_BINARY_DIR} EXCLUDE_FROM_ALL)
+  message(STATUS "Populated OpenCL ICD Loader")
+endif()
+list(APPEND INTEGRATED_OPENCL_INCLUDES ${OPENCL_ICD_LOADER_HEADERS_DIR})
+list(APPEND INTEGRATED_OPENCL_LIBRARIES ${opencl-icd-loader_BINARY_DIR}/Release/OpenCL.lib cfgmgr32.lib runtimeobject.lib)
+list(APPEND INTEGRATED_OPENCL_DEFINITIONS CL_TARGET_OPENCL_VERSION=120)
+
+# Build Independent Boost libraries
+include(ExternalProject)
+include(ProcessorCount)
+ProcessorCount(J)
+set(BOOST_BASE "${PROJECT_BINARY_DIR}/Boost")
+set(BOOST_BOOTSTRAP "${BOOST_BASE}/source/bootstrap.bat")
+set(BOOST_BUILD "${BOOST_BASE}/source/b2.exe")
+set(BOOST_FLAGS "")
+list(APPEND BOOST_SUBMODULES "libs/*" "tools/*")
```

Comment:
I guess we can workaround [1.5Gb+ size problem](https://github.com/microsoft/LightGBM/pull/3144#issuecomment-687654444) by specifying only required submodules here.

> IIRC the issue is that all boost libraries were being cloned even though only a handful were needed and I couldn't figure out a way to get around this.
  https://github.com/microsoft/LightGBM/pull/3144#issuecomment-689183648

Starting from `chrono, filesystem, headers, system, tools/*` by trials and errors I came up with the following list which allows to compile successfully, but I bet you know better which submodules are needed. Also, I believe we can reduce number of submodules under the `tools` folder.
```suggestion
list(APPEND BOOST_SUBMODULES "libs/algorithm" "libs/align" "libs/any" "libs/array" "libs/assert" "libs/bind" "libs/chrono" "libs/compute" "libs/concept_check" "libs/config" "libs/container" "libs/container_hash" "libs/core" "libs/detail" "libs/filesystem" "libs/foreach" "libs/format" "libs/function" "libs/function_types" "libs/fusion" "libs/headers" "libs/integer" "libs/io" "libs/iterator" "libs/lexical_cast" "libs/math" "libs/move" "libs/mpl" "libs/multi_index" "libs/numeric/conversion" "libs/optional" "libs/predef" "libs/preprocessor" "libs/property_tree" "libs/range" "libs/ratio" "libs/serialization" "libs/smart_ptr" "libs/static_assert" "libs/system" "libs/throw_exception" "libs/tuple" "libs/typeof" "libs/type_index" "libs/type_traits" "libs/utility" "libs/uuid" "libs/winapi" "tools/*")
```

I know, this list is looking super scary, but explicitly listing submodules it is possible to decrease build time and downloading size significantly (`1.5Gb+  ->  700Mb`)

Reply:
> Also, I believe we can reduce number of submodules under the tools folder.

OK, I've experimented more and it looks like we do not need submodules from `tools/` at all. I'm updating my yesterday original comment with new suggestion.

Reply:
@StrikerRUS , thank you for doing this exploration.

I'm not opposed to this change, and I've committed the suggested patch, but I think there are some considerations the maintainers should be aware of.  

One of the challenges in this patch was coaxing three different build systems (Boost's B2, cmake, and MSVC/make) to cooperate.  And as you know I was originally targeting more than just the Windows platform.  I found that the Boost dependency list differed significantly by platform -- on Linux (generally) only the few expected dependent submodules were built, on Windows (generally), many more submodules were fetched.   This was a bit surprising, but I didn't spend a lot of time chasing down why, because it seemed to me that downloading all submodules was always safe, and that even if I figured it all out for each platform, it would lead to complicated platform checking code that might turn out to be fragile in the long run (meaning a risk of not working anymore if the Boost libs or any of the three build systems changed or started interacting differently).

Of course I realize that there may be costs associated with fetching and processing code unecessarily in the CI pipeline, and also that non-Windows builds may never be needed, so I'm OK with the suggested change.  I just wanted to alert you all that you may have to rethink this targeted download approach if other platforms are added or the build becomes unstable.

Reply:
OK, I understand! Please correct me: as long as we fix a particular commit in the `boost` repository and building the library only for Windows users, it is safe to go with my list. In future we can update this list to support building on Linux machines or add conditioning `if(WIN32) ... else() ...` or fall back to downloading all submodules for all OSes.

Despite that we are using only free CI services (BTW, it will be not true anymore after introducing CUDA support #3160), we build nightly artifacts for each commit in `master` and in general I believe it is good practice to download things only that are needed.

Reply:
Not related to this PR, but just my curiosity. Don't you know were there any changes in macOS related to OpenCL support? Refer to https://github.com/microsoft/LightGBM/issues/1523#issuecomment-413357392 and #667. I believe some macOS users with eGPUs will benefit a lot.

Reply:
I'm sorry, but I'm not too familiar with the details of OpenCL on MacOS.  But the #667 issue you linked to seems (at first glance) like it could be resolved by an approach similar to this PR.  I'll look into it further when I get a bit of time.